text/0103-introduce-core-index-commitment.md

```diff
@@ -0,0 +1,240 @@
+
+### Reclaiming unused space in the descriptor
+
+The `CandidateDescriptor` currently includes `collator` and `signature` fields. The collator includes a signature on the following descriptor fields: parachain id, relay parent, validation data hash, validation code hash and the PoV hash.
+
+However, in practice, having a collator signature in the receipt on the relay chain does not provide any benefits as there is no mechanism to punish or reward collators that have provided bad parachain blocks.
+
+This proposal aims to remove the collator signature and all the logic that checks the collator signatures of candidate receipts. We use the first 7 reclaimed bytes to represent version, the core and session index, and fill the rest with zeroes. So, there is no change in the layout and length of the receipt. The new primitive is binary compatible with the old one.
+
+### UMP transport
+
+[CandidateCommitments](https://github.com/paritytech/polkadot-sdk/blob/b5029eb4fd6c7ffd8164b2fe12b71bad0c59c9f2/polkadot/primitives/src/v7/mod.rs#L682) remains unchanged as we will store scale encoded `UMPSignal` messages directly in the parachain UMP queue by outputing them in [upward_messages](https://github.com/paritytech/polkadot-sdk/blob/b5029eb4fd6c7ffd8164b2fe12b71bad0c59c9f2/polkadot/primitives/src/v7/mod.rs#L684). 
+
+The UMP queue layout is changed to allow the relay chain to receive both the XCM messages and `UMPSignal` messages. An empty message (empty `Vec<u8>`) is used to mark the end XCM messages and the start of `UMPSignal` messages.
+
+This way of representing the new messages has been chosen over introducing an enum wrapper to minimize breaking changes of XCM message decoding in tools like Subscan for example. 
+
+Example: 
+```rust
+[ XCM message1, XCM message2, ..., EMPTY message, UMPSignal::SelectCore ]
+```
+
+#### `UMPSignal` messages
+
+```rust
+/// An `u8` wrap around sequence number. Typically this would be the least significant byte of the parachain block number.
+pub struct CoreSelector(pub u8);
+
+/// An offset in the relay chain claim queue.
+pub struct ClaimQueueOffset(pub u8);
+
+/// Default claim queue offset
+pub const DEFAULT_CLAIM_QUEUE_OFFSET: ClaimQueueOffset = ClaimQueueOffset(1);
+
+pub enum UMPSignal {
+	/// A message sent by a parachain to select the core the candidate is commited to.
+	/// Relay chain validators, in particular backers, use the `CoreSelector` and `ClaimQueueOffset`
+	/// to compute the index of the core the candidate has commited to.
+	///
+	SelectCore(CoreSelector, ClaimQueueOffset),
+}
+```
+
+The parachain runtime is not concerned with the actual `CoreIndex` the candidate is intended for, 
+but must provide enough information for the validators and collators to compute it. `CoreSelector` 
+and `ClaimQueueOffset` use only 2 bytes and fullfil the requirement.
+
+**Example:**
+
+`cq_offset = 1` and `core_selector = 3`
+
+The table below represents a snapshot of the claim queue:
```

Comment:
I think we discussed this offline but I wanted to make sure this is clear.
The relay chain runtime cannot take into account the claim queue offset when backing a candidate. It can only back something if the claim reaches the top of the claim queue. By which time the claim at that offset could be something different. So the runtime must not use the claim queue offset provided in the inherent data. right?

A side effect is that that even if the claim queue offset used by the parachain is 1, a candidate can be backed sooner than that if it has a claim at offset 0. So this offset is really a max offset. But this sounds fine

Reply:
The claim queue offset is with regards to the used relay parent, hence it is fully deterministic. We also don't accept statements in the runtime other than for index 0, so we can just check the cores at position 0, if they match, we are good. The offset specification is only required for the backers to be able to deterministically check,